File: ainpc-core-plugin/src/main/java/ro/ainpc/world/MappingIndex.java
package ro.ainpc.world;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

class MappingIndex {

    private final Map<ChunkKey, List<WorldRegion>> regionsByChunk;
    private final Map<ChunkKey, List<WorldPlace>> placesByChunk;
    private final Map<ChunkKey, List<WorldNode>> nodesByChunk;

    MappingIndex() {
        this.regionsByChunk = new HashMap<>();
        this.placesByChunk = new HashMap<>();
        this.nodesByChunk = new HashMap<>();
    }

    void clear() {
        regionsByChunk.clear();
        placesByChunk.clear();
        nodesByChunk.clear();
    }

    void indexRegion(WorldRegion region) {
        forEachChunk(region.getWorldName(), region.getMinX(), region.getMinZ(), region.getMaxX(), region.getMaxZ(),
            key -> regionsByChunk.computeIfAbsent(key, ignored -> new ArrayList<>()).add(region));
    }

    void indexPlace(WorldPlace place) {
        forEachChunk(place.getWorldName(), place.getMinX(), place.getMinZ(), place.getMaxX(), place.getMaxZ(),
            key -> placesByChunk.computeIfAbsent(key, ignored -> new ArrayList<>()).add(place));
    }

    void indexNode(WorldNode node) {
        int minX = (int) Math.floor(node.getX() - node.getRadius());
        int minZ = (int) Math.floor(node.getZ() - node.getRadius());
        int maxX = (int) Math.ceil(node.getX() + node.getRadius());
        int maxZ = (int) Math.ceil(node.getZ() + node.getRadius());
        forEachChunk(node.getWorldName(), minX, minZ, maxX, maxZ,
            key -> nodesByChunk.computeIfAbsent(key, ignored -> new ArrayList<>()).add(node));
    }

    WorldRegion findRegion(String worldName, int x, int y, int z) {
        for (WorldRegion region : regionsByChunk.getOrDefault(keyAt(worldName, x, z), List.of())) {
            if (region.contains(worldName, x, y, z)) {
                return region;
            }
        }
        return null;
    }

    WorldPlace findPlace(String worldName, int x, int y, int z) {
        for (WorldPlace place : placesByChunk.getOrDefault(keyAt(worldName, x, z), List.of())) {
            if (place.contains(worldName, x, y, z)) {
                return place;
            }
        }
        return null;
    }

    WorldNode findNode(String worldName, double x, double y, double z) {
        for (WorldNode node : nodesByChunk.getOrDefault(keyAt(worldName, (int) Math.floor(x), (int) Math.floor(z)), List.of())) {
            if (node.contains(worldName, x, y, z)) {
                return node;
            }
        }
        return null;
    }

    List<WorldNode> findNodesNear(String worldName, double x, double y, double z, double radius, int limit) {
        if (worldName == null || radius < 0.0 || limit == 0) {
            return List.of();
        }

        Set<WorldNode> candidates = new LinkedHashSet<>();
        int minX = (int) Math.floor(x - radius);
        int minZ = (int) Math.floor(z - radius);
        int maxX = (int) Math.ceil(x + radius);
        int maxZ = (int) Math.ceil(z + radius);
        forEachChunk(worldName, minX, minZ, maxX, maxZ,
            key -> candidates.addAll(nodesByChunk.getOrDefault(key, List.of())));

        return candidates.stream()
            .filter(node -> node.isNear(worldName, x, y, z, radius + node.getRadius()))
            .sorted(Comparator.comparingDouble(node -> node.distanceSquared(worldName, x, y, z)))
            .limit(limit > 0 ? limit : Long.MAX_VALUE)
            .toList();
    }

    int indexedRegionChunks() {
        return regionsByChunk.size();
    }

    int indexedPlaceChunks() {
        return placesByChunk.size();
    }

    int indexedNodeChunks() {
        return nodesByChunk.size();
    }

    private void forEachChunk(String worldName,
                              int minX,
                              int minZ,
                              int maxX,
                              int maxZ,
                              ChunkConsumer consumer) {
        int minChunkX = toChunk(minX);
        int maxChunkX = toChunk(maxX);
        int minChunkZ = toChunk(minZ);
        int maxChunkZ = toChunk(maxZ);

        for (int chunkX = minChunkX; chunkX <= maxChunkX; chunkX++) {
            for (int chunkZ = minChunkZ; chunkZ <= maxChunkZ; chunkZ++) {
                consumer.accept(new ChunkKey(normalizeWorld(worldName), chunkX, chunkZ));
            }
        }
    }

    private ChunkKey keyAt(String worldName, int x, int z) {
        return new ChunkKey(normalizeWorld(worldName), toChunk(x), toChunk(z));
    }

    private int toChunk(int blockCoordinate) {
        return Math.floorDiv(blockCoordinate, 16);
    }

    private String normalizeWorld(String worldName) {
        return worldName == null ? "" : worldName.toLowerCase(Locale.ROOT);
    }

    private interface ChunkConsumer {
        void accept(ChunkKey key);
    }

    private record ChunkKey(String worldName, int chunkX, int chunkZ) {
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/world/MappingIndex.kt
package ro.ainpc.world

import java.util.LinkedHashSet
import java.util.Locale
import kotlin.math.ceil
import kotlin.math.floor

internal class MappingIndex {
    private val regionsByChunk: MutableMap<ChunkKey, MutableList<WorldRegion>> = HashMap()
    private val placesByChunk: MutableMap<ChunkKey, MutableList<WorldPlace>> = HashMap()
    private val nodesByChunk: MutableMap<ChunkKey, MutableList<WorldNode>> = HashMap()

    fun clear() {
        regionsByChunk.clear()
        placesByChunk.clear()
        nodesByChunk.clear()
    }

    fun indexRegion(region: WorldRegion) {
        forEachChunk(region.worldName, region.minX, region.minZ, region.maxX, region.maxZ) { key ->
            regionsByChunk.computeIfAbsent(key) { ArrayList() }.add(region)
        }
    }

    fun indexPlace(place: WorldPlace) {
        forEachChunk(place.worldName, place.minX, place.minZ, place.maxX, place.maxZ) { key ->
            placesByChunk.computeIfAbsent(key) { ArrayList() }.add(place)
        }
    }

    fun indexNode(node: WorldNode) {
        val minX = floor(node.x - node.radius).toInt()
        val minZ = floor(node.z - node.radius).toInt()
        val maxX = ceil(node.x + node.radius).toInt()
        val maxZ = ceil(node.z + node.radius).toInt()
        forEachChunk(node.worldName, minX, minZ, maxX, maxZ) { key ->
            nodesByChunk.computeIfAbsent(key) { ArrayList() }.add(node)
        }
    }

    fun findRegion(worldName: String?, x: Int, y: Int, z: Int): WorldRegion? {
        for (region in regionsByChunk[keyAt(worldName, x, z)] ?: emptyList()) {
            if (region.contains(worldName, x, y, z)) {
                return region
            }
        }
        return null
    }

    fun findPlace(worldName: String?, x: Int, y: Int, z: Int): WorldPlace? {
        for (place in placesByChunk[keyAt(worldName, x, z)] ?: emptyList()) {
            if (place.contains(worldName, x, y, z)) {
                return place
            }
        }
        return null
    }

    fun findNode(worldName: String?, x: Double, y: Double, z: Double): WorldNode? {
        val key = keyAt(worldName, floor(x).toInt(), floor(z).toInt())
        for (node in nodesByChunk[key] ?: emptyList()) {
            if (node.contains(worldName, x, y, z)) {
                return node
            }
        }
        return null
    }

    fun findNodesNear(worldName: String?, x: Double, y: Double, z: Double, radius: Double, limit: Int): List<WorldNode> {
        if (worldName == null || radius < 0.0 || limit == 0) {
            return emptyList()
        }

        val candidates = LinkedHashSet<WorldNode>()
        val minX = floor(x - radius).toInt()
        val minZ = floor(z - radius).toInt()
        val maxX = ceil(x + radius).toInt()
        val maxZ = ceil(z + radius).toInt()
        forEachChunk(worldName, minX, minZ, maxX, maxZ) { key ->
            candidates.addAll(nodesByChunk[key] ?: emptyList())
        }

        return candidates.asSequence()
            .filter { node -> node.isNear(worldName, x, y, z, radius + node.radius) }
            .sortedBy { node -> node.distanceSquared(worldName, x, y, z) }
            .let { sequence -> if (limit > 0) sequence.take(limit) else sequence }
            .toList()
    }

    fun indexedRegionChunks(): Int = regionsByChunk.size

    fun indexedPlaceChunks(): Int = placesByChunk.size

    fun indexedNodeChunks(): Int = nodesByChunk.size

    private fun forEachChunk(
        worldName: String?,
        minX: Int,
        minZ: Int,
        maxX: Int,
        maxZ: Int,
        consumer: (ChunkKey) -> Unit
    ) {
        val minChunkX = toChunk(minX)
        val maxChunkX = toChunk(maxX)
        val minChunkZ = toChunk(minZ)
        val maxChunkZ = toChunk(maxZ)

        for (chunkX in minChunkX..maxChunkX) {
            for (chunkZ in minChunkZ..maxChunkZ) {
                consumer(ChunkKey(normalizeWorld(worldName), chunkX, chunkZ))
            }
        }
    }

    private fun keyAt(worldName: String?, x: Int, z: Int): ChunkKey {
        return ChunkKey(normalizeWorld(worldName), toChunk(x), toChunk(z))
    }

    private fun toChunk(blockCoordinate: Int): Int = Math.floorDiv(blockCoordinate, 16)

    private fun normalizeWorld(worldName: String?): String {
        return worldName?.lowercase(Locale.ROOT).orEmpty()
    }

    private data class ChunkKey(
        val worldName: String,
        val chunkX: Int,
        val chunkZ: Int
    )
}
